### PR TITLE
Back

### PR DESCRIPTION
**Pull request by GitBluub**
have fun

### CODE REVIEW

File: api/sql/dump.sql [hunk 1]
```diff
@@ -2,14 +2,19 @@ CREATE TABLE IF NOT EXISTS users (
...
+	external_tokens JSONB[] NOT NULL
 );
 
 CREATE TABLE IF NOT EXISTS pipelines (
 	id SERIAL PRIMARY KEY,
 	name VARCHAR(255) NOT NULL,
 	type VARCHAR(255) NOT NULL,
-	params JSONB NOT NULL
+	params JSONB NOT NULL,
+  user_id INTEGER NOT NULL,
```
Comment (AnonymusRaccoon) [inline api/sql/dump.sql:14]: ah

File: api/src/Api/OIDC.hs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import Servant.API.Generic (type (:-))
+import Servant.Server.Generic (AsServerT)
+
+oauthHandler :: Service -> Maybe String -> AppM NoContent
+oauthHandler service (Just code) = do
+    tokens <- liftIO $ getOauthTokens service code
+    case tokens of
+        Nothing -> throwError err400
+        Just t -> do
+            updateTokens (UserId 1) t
```
Comment (AnonymusRaccoon) [inline api/src/Api/OIDC.hs:23]: Nice user,

File: api/src/Api/Pipeline.hs [hunk 3]
```diff
@@ -54,43 +56,52 @@ $(deriveJSON defaultOptions ''ReactionData)
...
     reactions <- getReactionsByPipelineId' pipelineId
     let actionResult = PipelineData (pipelineName pipeline) (pipelineType pipeline) (pipelineParams pipeline)
-    let reactionsResult = fmap (\x -> ReactionData (reactionType x) (reactionParams x)) reactions 
+    let reactionsResult = fmap (\x -> ReactionData (reactionType x) (reactionParams x)) reactions
     return $ PostPipelineData actionResult reactionsResult
 
 postPipelineHandler :: PostPipelineData -> AppM [ReactionId]
 postPipelineHandler x = do
-    actionId <- createPipeline $ Pipeline (PipelineId 1) (pipelineDataName p) (pipelineDataType p) (pipelineDataParams p)
+    actionId <- createPipeline $ Pipeline (PipelineId 1) (name p) (pType p) (pParams p) (UserId 1)
```
Comment (AnonymusRaccoon) [inline api/src/Api/Pipeline.hs:77]: Nice id

Comment (AnonymusRaccoon) [reply api/src/Api/Pipeline.hs:77]: Ok nvm it is not used.

Comment (AnonymusRaccoon) [inline api/src/Api/Pipeline.hs:77]: Nice user id

File: api/src/Api/Pipeline.hs [hunk 4]
```diff
@@ -54,43 +56,52 @@ $(deriveJSON defaultOptions ''ReactionData)
...
     sequence $ mapInd (reactionMap (head actionId)) r
```
Comment (AnonymusRaccoon) [inline api/src/Api/Pipeline.hs:78]: You should place your head in the createPipeline and not here.

File: api/src/Api/Pipeline.hs [hunk 5]
```diff
@@ -54,43 +56,52 @@ $(deriveJSON defaultOptions ''ReactionData)
...
-        reactionMap actionId s i = do
-            res <- createReaction $ Reaction (ReactionId 1) (reactionDataType s) (reactionDataParams s) actionId (fromIntegral i)
-            return $ head res
+  where
+    p = action x
+    r = reactions x
+    reactionMap :: PipelineId -> ReactionData -> Int -> AppM ReactionId
+    reactionMap actionId s i = do
+        res <- createReaction $ Reaction (ReactionId 1) (rType s) (rParams s) actionId (fromIntegral i)
+        return $ head res
```
Comment (AnonymusRaccoon) [inline api/src/Api/Pipeline.hs:85]: Same as createPipeline, move the head

File: api/src/Db/Reaction.hs [hunk 6]
```diff
@@ -1,74 +1,84 @@
...
+    FromJSON,
+    ToJSON,
+    defaultOptions,
+    eitherDecode,
+ )
+import Data.Aeson.TH (deriveJSON)
 import Data.Int (Int64)
+import Data.Text (Text)
+import GHC.Generics (Generic)
+import Rel8 (Column, DBEq, DBType, Expr, Insert (Insert, returning), JSONBEncoded (JSONBEncoded), Name, OnConflict (DoNothing), Query, ReadShow (ReadShow), Rel8able, Result, Returning (Projection), TableSchema (TableSchema, columns, name, schema), each, into, lit, nextval, onConflict, returning, rows, unsafeCastExpr, values, where_, (==.))
```
Comment (AnonymusRaccoon) [inline api/src/Db/Reaction.hs:22]: Nice line.

File: api/src/Db/User.hs [hunk 7]
```diff
@@ -3,31 +3,34 @@
...
 import Data.Text (Text)
-import Data.Aeson (FromJSON, ToJSON)
+import GHC.Generics (Generic)
 import Password (HashedPassword)
+import Servant.Auth.JWT
 
 import Core.User
-import Rel8 (Column, Rel8able, Result, TableSchema (TableSchema, schema, name, columns), Name, Query, Expr, Insert (Insert, returning), returning, onConflict, rows, into, where_, (==.), lit, values, unsafeCastExpr, nextval, OnConflict (DoNothing), Returning (Projection), each)
+import Data.List (findIndex)
+import Rel8 (Column, Expr, Insert (Insert, returning), Name, OnConflict (DoNothing), Query, Rel8able, Result, Returning (NumberOfRowsAffected, Projection), TableSchema (TableSchema, columns, name, schema), Update (Update, from, returning, set, target, updateWhere), each, into, lit, nextval, onConflict, returning, rows, unsafeCastExpr, values, where_, (==.))
```
Comment (AnonymusRaccoon) [inline api/src/Db/User.hs:24]: I'll give up

**Comment by GitBluub** [comment]
JWT are not read for the workflows and oauth routes so users are hardcoded to the first one

**Pull request merged**